activation/wire/wire_v1.go

```diff
@@ -132,9 +132,11 @@ func (atx *ActivationTxV1) SignedBytes() []byte {
 }
 
 func (atx *ActivationTxV1) HashInnerBytes() (result types.Hash32) {
-	h := hash.New()
+	h := hash.GetHasher()
 	codec.MustEncodeTo(h, &atx.InnerActivationTxV1)
 	h.Sum(result[:0])
+	h.Reset()
+	hash.PutHasher(h)
```

Comment:
Forgetting to `Reset()` the hasher would have tragic consequences. How about resetting it automatically in `PutHasher()`?

Reply:
Thanks @poszu for the review. So this came up obviously while sketching this change out and here is the reasoning on why I've done it this way:
- hashing is a fairly easy operation which is usually compact in terms of usage, so it is easy to see where the hasher is acquired and therefore make sure it is not held too long (so it goes back into the pool ASAP) and release it manually
- The reason why I kept it simply in the pool is to just make sure objects come in and out of the pool, without too much boilerplate around it
- The methods are documented as requiring the consumers to do a `Reset`, just as in some APIs it is documented that consumers are expected to hold a lock on some mutex or synchronization primitive
- I did not put the cleanups in a `defer` statement because putting it in a closure means it adds another minimum of 2 operations for every defer call (one of evaluating the empty closure at the `defer` call, and another when the anonymous function that exists inside of it). Obviously this can be replaced by just calling `defer` twice with the `Reset` and `PutHasher` separately. Since we're doing a lot of hashing I would try to reduce overhead on all of these corollary ops.
- If you insist of having the cleanups in a `defer` statement (usually I would vote to do the same, apart from several exceptions), I would suggest to change the `PutHasher` method to be `PutResetHasher` instead, and that the call to `Reset` would then be inlined there, and then call it in a single `defer` call in all places where it could be used as a `defer`. There are a few places in the PR where I wouldn't use a defer because it would mean that the instance unnecessarily stays with the consuming component and then I would just release it manually so that other callers could also make use it ASAP. Let me know if this is acceptable

Reply:
> - The methods are documented as requiring the consumers to do a Reset, just as in some APIs it is documented that consumers are expected to hold a lock on some mutex or synchronization primitive

I know, but there is nothing enforcing this rule and if it can be forgotten, it will be forgotten. IMHO, if the hasher MUST be reset before it is returned to the pool, then it should be implemented in a way to make it impossible to make a mistake. The easiest way of achieving it is to reset it automatically in the `Put` function.

>I would suggest to change the PutHasher method to be PutResetHasher

I think that's not necessary. The user of this API doesn't need to know that the hasher is reset. The user only wants to get a hasher and return it.

> I did not put the cleanups in a defer statement because putting it in a closure means it adds another minimum of 2 operations for every defer call (one of evaluating the empty closure at the defer call, and another when the anonymous function that exists inside of it). 

I don't understand, do you mean that:
```go 
defer func() {
    h.Reset()
    hash.Put(h)
}()
```
is too much overhead because of the extra closure? I didn't check but I would be surprised if that's not inlined.

>There are a few places in the PR where I wouldn't use a defer because it would mean that the instance unnecessarily stays with the consuming component and then I would just release it manually so that other callers could also make use it ASAP.

I understand, then it's good to release it ASAP manually as you say. The alternative could be to extract the hashing part into a function and continue relying on defer there.

Reply:
Thanks @poszu I'll amend as suggested.

And as you said the defer closure will indeed get inlined (I wrote this simple example to see for my own eyes): https://termbin.com/umrn